packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -451,15 +472,100 @@ export class DbTargetGlobal {
      * until the backend has caught up to this snapshot version again. This
      * prevents our cache from ever going backwards in time.
      */
-    public lastRemoteSnapshotVersion: DbTimestamp
+    public lastRemoteSnapshotVersion: DbTimestamp,
+    /**
+     * A cache of the number of targets persisted.
+     */
+    public targetCount: number
   ) {}
 }
 
+function createQueryCache(db: IDBDatabase): void {
+  const targetDocumentsStore = db.createObjectStore(DbTargetDocument.store, {
+    keyPath: DbTargetDocument.keyPath as KeyPath
+  });
+  targetDocumentsStore.createIndex(
+    DbTargetDocument.documentTargetsIndex,
+    DbTargetDocument.documentTargetsKeyPath,
+    { unique: true }
+  );
+
+  const targetStore = db.createObjectStore(DbTarget.store, {
+    keyPath: DbTarget.keyPath
+  });
+
+  // NOTE: This is unique only because the TargetId is the suffix.
+  targetStore.createIndex(
+    DbTarget.queryTargetsIndexName,
+    DbTarget.queryTargetsKeyPath,
+    { unique: true }
+  );
+  db.createObjectStore(DbTargetGlobal.store);
+}
+
 /**
- * The list of all IndexedDB stored used by the SDK. This is used when creating
- * transactions so that access across all stores is done atomically.
+ * Counts the number of targets persisted and adds that value to the target
+ * global singleton.
+ * @param {PersistencePromise<void>} p The promise used to keep the transaction open
+ * @param {IDBTransaction} txn The version upgrade transaction for indexeddb
+ * @return {PersistencePromise<void>} A promise of the operations to add the count to the metadata row.
  */
-export const ALL_STORES = [
+function addTargetCount(
+  p: PersistencePromise<void>,
+  txn: IDBTransaction
+): PersistencePromise<void> {
+  const globalStore = txn.objectStore(DbTargetGlobal.store);
+  return p
+    .next(() => wrapRequest<number>(txn.objectStore(DbTarget.store).count()))
+    .next((count: number) =>
+      wrapRequest<DbTargetGlobal>(globalStore.get(DbTargetGlobal.key)).next(
+        (metadata: DbTargetGlobal) => {
+          metadata.targetCount = count;
+          return metadata;
+        }
+      )
+    )
+    .next((metadata: DbTargetGlobal) =>
+      wrapRequest<void>(globalStore.put(metadata, DbTargetGlobal.key))
```

Comment:
You can combine these two Promise branches.

Reply:
Combined a little.